### PR TITLE
update tokencache serialization api so it's only visible during notif…

### PR DESCRIPTION
**Pull request by MarkZuber**
…ications

### CODE REVIEW

**Comment by henrik-me** [comment]
failing integration tests?

File: src/Microsoft.Identity.Client/ITokenCache.cs [hunk 1]
```diff
@@ -89,109 +89,109 @@ public interface ITokenCache
...
+        ///// Deserializes the token cache to the ADAL.NET 3.x cache format.
+        ///// If you need to maintain SSO between an application using ADAL 3.x or MSAL 2.x and this application using MSAL 3.x,
+        ///// you might also want to serialize and deserialize with <see cref="SerializeAdalV3"/>/<see cref="DeserializeAdalV3"/> or <see cref="SerializeMsalV2"/>/<see cref="DeserializeMsalV2"/>,
+        ///// otherwise just use <see cref="SerializeMsalV3"/>/<see cref="DeserializeMsalV3"/>.
+        ///// </summary>
+        ///// <param name="adalV3State">Byte stream representation of the cache</param>
+        ///// <remarks>
+        ///// <see cref="SerializeMsalV3"/>/<see cref="DeserializeMsalV3"/> is compatible with other MSAL libraries such as MSAL for Python and MSAL for Java.
+        ///// </remarks>
+        //void DeserializeAdalV3(byte[] adalV3State);
```
Comment (henrik-me) [inline src/Microsoft.Identity.Client/ITokenCache.cs]: [obsolete] comments should probably be something like what you have in the remarks for SetBeforeAccessAsync and SetAfterAccessAsync ?
(applies to all commented out methods)

**Comment by trwalke** [comment]
@henrik-me integration tests are failing due to a recent update to the test infrastructure that introduced instability. (Selenium driver is timing out). Pleas try re running @MarkZuber

File: src/Microsoft.Identity.Client/MigrationAid/TokenCache.MigrationAid.cs [hunk 2]
```diff
@@ -78,38 +78,38 @@ public void DeserializeUnifiedAndAdalCache(CacheData cacheData)
...
+        /// <see cref="ITokenCacheSerializer.SerializeMsalV3"/>/<see cref="ITokenCacheSerializer.DeserializeMsalV3"/> Is our recommended way of serializing/deserializing.
+        /// <see cref="ITokenCacheSerializer.DeserializeAdalV3"/> For interoperability with ADAL.NET v3
         /// </summary>
         /// <param name="msalV2State">Array of bytes containing serialized MSAL.NET V2 cache data</param>
         /// <remarks>
-        /// <see cref="SerializeMsalV3"/>/<see cref="DeserializeMsalV3"/> is compatible with other MSAL libraries such as MSAL for Python and MSAL for Java.
+        /// <see cref="ITokenCacheSerializer.SerializeMsalV3"/>/<see cref="ITokenCacheSerializer.DeserializeMsalV3"/> is compatible with other MSAL libraries such as MSAL for Python and MSAL for Java.
         /// <paramref name="msalV2State"/>Is a Json blob containing access tokens, refresh tokens, id tokens and accounts information.
         /// </remarks>
         [Obsolete("This is expected to be removed in MSAL.NET v4 and ADAL.NET v6. We recommend using SerializeMsalV3/DeserializeMsalV3. Read more: https://aka.ms/msal-net-3x-cache-breaking-change", false)]
```
Comment (henrik-me) [inline src/Microsoft.Identity.Client/MigrationAid/TokenCache.MigrationAid.cs]: >expected [](start = 27, length = 8)

same elsewhere: This is removed in MSAL.NET v4 and expected to be removed in ADAL.NET v6. ...

File: src/Microsoft.Identity.Client/ITokenCache.cs (hunk 1 above)
Comment (MarkZuber) [reply src/Microsoft.Identity.Client/ITokenCache.cs]: Agreed.  I submitted this to get a test build other than on my machine.  VS 16.1 borked my ability to fully build locally.  I think the integration tests are broken due to Bogdan saying he added some traces that were hanging the tests.  But I will investigate further as well.

File: src/Microsoft.Identity.Client/ITokenCache.cs [hunk 3]
```diff
@@ -89,109 +89,109 @@ public interface ITokenCache
...
+
+        /// <summary>
+        /// Functionality replaced by <see cref="ITokenCacheSerializer.SerializeMsalV2"/>. See https://aka.ms/msal-net-3x-cache-breaking-change
```
Comment (jmprieur) [inline src/Microsoft.Identity.Client/ITokenCache.cs]: I love this!
The aka.ms link should probably be https://aka.ms/msal-net-4x-cache-breaking-change

File: src/Microsoft.Identity.Client/ITokenCache.cs [hunk 4]
```diff
@@ -89,109 +89,109 @@ public interface ITokenCache
...
-        /// Functionality replaced by <see cref="DeserializeMsalV2"/>.  See https://aka.ms/msal-net-3x-cache-breaking-change
+        /// Functionality replaced by <see cref="ITokenCacheSerializer.DeserializeMsalV2"/>.  See https://aka.ms/msal-net-3x-cache-breaking-change
         /// </summary>
         /// <param name="msalV2State"></param>
         [Obsolete("This is expected to be removed in MSAL.NET v3 and ADAL.NET v5. We recommend using SerializeMsalV3/DeserializeMsalV3. Read more: https://aka.ms/msal-net-3x-cache-breaking-change", false)]
         void Deserialize(byte[] msalV2State);
 
         /// <summary>
-        /// Functionality replaced by <see cref="SerializeMsalV2"/> and <see cref="SerializeAdalV3"/>
+        /// Functionality replaced by <see cref="ITokenCacheSerializer.SerializeMsalV2"/> and <see cref="ITokenCacheSerializer.SerializeAdalV3"/>
```
Comment (jmprieur) [inline src/Microsoft.Identity.Client/ITokenCache.cs]: https://aka.ms/msal-net-4x-cache-breaking-change

File: src/Microsoft.Identity.Client/ITokenCache.cs [hunk 5]
```diff
@@ -89,109 +89,109 @@ public interface ITokenCache
...
         /// See https://aka.ms/msal-net-3x-cache-breaking-change
         /// </summary>
         /// <returns></returns>
         [Obsolete("This is expected to be removed in MSAL.NET v3 and ADAL.NET v5. We recommend using SerializeMsalV3/DeserializeMsalV3. Read more: https://aka.ms/msal-net-3x-cache-breaking-change", false)]
         CacheData SerializeUnifiedAndAdalCache();
 
         /// <summary>
-        /// Functionality replaced by <see cref="DeserializeMsalV2"/> and <see cref="DeserializeAdalV3"/>
+        /// Functionality replaced by <see cref="ITokenCacheSerializer.DeserializeMsalV2"/> and <see cref="ITokenCacheSerializer.DeserializeAdalV3"/>
```
Comment (jmprieur) [inline src/Microsoft.Identity.Client/ITokenCache.cs]: https://aka.ms/msal-net-4x-cache-breaking-change

File: src/Microsoft.Identity.Client/MigrationAid/TokenCache.MigrationAid.cs [hunk 6]
```diff
@@ -78,38 +78,38 @@ public void DeserializeUnifiedAndAdalCache(CacheData cacheData)
...
+        /// <see cref="ITokenCacheSerializer.SerializeMsalV3"/>/<see cref="ITokenCacheSerializer.DeserializeMsalV3"/> Is our recommended way of serializing/deserializing.
+        /// <see cref="ITokenCacheSerializer.SerializeAdalV3"/> For interoperability with ADAL.NET v3.
         /// </summary>
-        /// <returns>array of bytes, <see cref="SerializeMsalV2"/></returns>
+        /// <returns>array of bytes, <see cref="ITokenCacheSerializer.SerializeMsalV2"/></returns>
         /// <remarks>
-        /// <see cref="SerializeMsalV3"/>/<see cref="DeserializeMsalV3"/> is compatible with other MSAL libraries such as MSAL for Python and MSAL for Java.
+        /// <see cref="ITokenCacheSerializer.SerializeMsalV3"/>/<see cref="ITokenCacheSerializer.DeserializeMsalV3"/> is compatible with other MSAL libraries such as MSAL for Python and MSAL for Java.
         /// </remarks>
         [Obsolete("This is expected to be removed in MSAL.NET v4 and ADAL.NET v6. We recommend using SerializeMsalV3/DeserializeMsalV3. Read more: https://aka.ms/msal-net-3x-cache-breaking-change", false)]
```
Comment (jmprieur) [inline src/Microsoft.Identity.Client/MigrationAid/TokenCache.MigrationAid.cs]: this is removed (as we now are in MSAL.NET v4)?
Now a link to https://aka.ms/msal-net-4x-cache-breaking-change?

File: src/Microsoft.Identity.Client/TokenCache.Notifications.cs [hunk 7]
```diff
@@ -137,5 +137,15 @@ public void SetBeforeWriteAsync(Func<TokenCacheNotificationArgs, Task> beforeWri
             AsyncBeforeWrite = beforeWrite;
         }
 #endif
+
+        private static void GuardOnMobilePlatforms()
+        {
+#if ANDROID || iOS
+        throw new PlatformNotSupportedException("You should not use these TokenCache methods object on mobile platforms. " +
+            "They meant to allow applications to define their own storage strategy on .net desktop and non-mobile platforms such as .net core. " +
+            "On mobile platforms, a secure and performant storage mechanism is implemeted by MSAL. " +
```
Comment (jmprieur) [inline src/Microsoft.Identity.Client/TokenCache.Notifications.cs]: typo: implemeted => implemented.

Maybe use the active voice: "On mobile platforms, MSAL.NET implements already a secure and performant storage mechanism."

File: src/Microsoft.Identity.Client/TokenCacheNotificationArgs.cs [hunk 8]
```diff
@@ -5,25 +5,37 @@ namespace Microsoft.Identity.Client
...
+            Account = account;
+            HasStateChanged = hasStateChanged;
+        }
+
         /// <summary>
-        /// Gets the <see cref="ITokenCache"/> involved in the transaction
+        /// Gets the <see cref="ITokenCacheSerializer"/> involved in the transaction
         /// </summary>
-        public ITokenCache TokenCache { get; internal set; }
+        public ITokenCacheSerializer TokenCache { get; }
```
Comment (jmprieur) [inline src/Microsoft.Identity.Client/TokenCacheNotificationArgs.cs]: I guess we used the same name to avoid people to have to change a lot of code?

Comment (MarkZuber) [reply src/Microsoft.Identity.Client/TokenCacheNotificationArgs.cs]: Yes, that's correct.  For people who already are serializing in their callbacks, this fix will "just work" which is ideally most of the cases.

Comment (jmprieur) [reply src/Microsoft.Identity.Client/TokenCacheNotificationArgs.cs]: Thanks for confirming. Awesome. The breaking change is really making sure that people do the right thing.

File: src/Microsoft.Identity.Client/TokenCache.Notifications.cs [hunk 9]
```diff
@@ -137,5 +137,15 @@ public void SetBeforeWriteAsync(Func<TokenCacheNotificationArgs, Task> beforeWri
             AsyncBeforeWrite = beforeWrite;
         }
 #endif
+
+        private static void GuardOnMobilePlatforms()
+        {
+#if ANDROID || iOS
+        throw new PlatformNotSupportedException("You should not use these TokenCache methods object on mobile platforms. " +
+            "They meant to allow applications to define their own storage strategy on .net desktop and non-mobile platforms such as .net core. " +
```
Comment (bgavrilMS) [inline src/Microsoft.Identity.Client/TokenCache.Notifications.cs]: Typo "they are meant"

File: tests/devapps/XForms/XForms.Android/Properties/AndroidManifest.xml [hunk 10]
```diff
@@ -1,6 +1,6 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <manifest xmlns:android="http://schemas.android.com/apk/res/android" android:installLocation="auto" package="com.Microsoft.XFormsDroid.MSAL">
```
Comment (bgavrilMS) [inline tests/devapps/XForms/XForms.Android/Properties/AndroidManifest.xml]: There should be a lot of devapps that need changing.

File: src/Microsoft.Identity.Client/TokenCache.Notifications.cs [hunk 11]
```diff
@@ -137,5 +137,15 @@ public void SetBeforeWriteAsync(Func<TokenCacheNotificationArgs, Task> beforeWri
             AsyncBeforeWrite = beforeWrite;
         }
 #endif
+
+        private static void GuardOnMobilePlatforms()
+        {
+#if ANDROID || iOS
+        throw new PlatformNotSupportedException("You should not use these TokenCache methods object on mobile platforms. " +
```
Comment (jennyf19) [inline src/Microsoft.Identity.Client/TokenCache.Notifications.cs]: >object [](start = 93, length = 6)

remove

File: src/Microsoft.Identity.Client/TokenCache.Notifications.cs (hunk 7 above)
Comment (jennyf19) [reply src/Microsoft.Identity.Client/TokenCache.Notifications.cs]: +1

---
In reply to: [286805957](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1170#discussion_r286805957) [](ancestors = 286805957)

File: src/Microsoft.Identity.Client/TokenCacheNotificationArgs.cs (hunk 8 above)
Comment (jennyf19) [reply src/Microsoft.Identity.Client/TokenCacheNotificationArgs.cs]: nice.

---
In reply to: [287077505](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1170#discussion_r287077505) [](ancestors = 287077505)

**Pull request merged**